<a href="https://colab.research.google.com/github/omkar123katare/EDA-hotel_bookings/blob/main/EDA_hotel_bookings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **STEP 1 : Dataset Cleaning**


# **Mounting google drive**


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# **Importing important libraries for EDA**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

#**Reading the dataset from drive**

In [ ]:
file_path='/content/drive/MyDrive/Copy of Hotel Bookings.csv'
hotel_bookings=pd.read_csv(file_path)

#**First 5 entries of the dataset**

In [ ]:
hotel_bookings.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 32 columns]

# **Last 5 entries of the dataset**

In [ ]:
hotel_bookings.tail()

hotel  is_canceled  ...  reservation_status  reservation_status_date
119385  City Hotel            0  ...           Check-Out               2017-09-06
119386  City Hotel            0  ...           Check-Out               2017-09-07
119387  City Hotel            0  ...           Check-Out               2017-09-07
119388  City Hotel            0  ...           Check-Out               2017-09-07
119389  City Hotel            0  ...           Check-Out               2017-09-07

[5 rows x 32 columns]

#**Shape of dataset**

In [ ]:
hotel_bookings.shape

(119390, 32)

#**Dataset information- Column wise Number of non-Null values and Datatypet**

In [ ]:
hotel_bookings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

#**Summary of the data- columnwise**

In [ ]:
# mean, maximum, minimum values of numberical columns.
hotel_bookings.describe()

is_canceled  ...  total_of_special_requests
count  119390.000000  ...              119390.000000
mean        0.370416  ...                   0.571363
std         0.482918  ...                   0.792798
min         0.000000  ...                   0.000000
25%         0.000000  ...                   0.000000
50%         0.000000  ...                   0.000000
75%         1.000000  ...                   1.000000
max         1.000000  ...                   5.000000

[8 rows x 20 columns]

#**Names of Columns in the dataset.**

In [ ]:
list(hotel_bookings.columns)

['hotel',
 'is_canceled',
 'lead_time',
 'arrival_date_year',
 'arrival_date_month',
 'arrival_date_week_number',
 'arrival_date_day_of_month',
 'stays_in_weekend_nights',
 'stays_in_week_nights',
 'adults',
 'children',
 'babies',
 'meal',
 'country',
 'market_segment',
 'distribution_channel',
 'is_repeated_guest',
 'previous_cancellations',
 'previous_bookings_not_canceled',
 'reserved_room_type',
 'assigned_room_type',
 'booking_changes',
 'deposit_type',
 'agent',
 'company',
 'days_in_waiting_list',
 'customer_type',
 'adr',
 'required_car_parking_spaces',
 'total_of_special_requests',
 'reservation_status',
 'reservation_status_date']

#**Handling null values**

In [ ]:
hotel_bookings.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         


1. There are 4 null values in 'children' column.
2. There are 488 null values in 'country' column.
3. There are 16340 null values in 'agent' column.
4. There are 112593 null values in 'company' column 

In [ ]:
# 'agent' & 'company' are two columns with huge amount of null values in it.

new_hotel_df = hotel_bookings.drop(['agent','company'],axis = 1)
new_hotel_df.head()

# We hve dropped 'agent' and 'company' columns.

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 30 columns]

**Values in adults, childrens and babies columns can't be zero at the same time**

In [ ]:
# so now we are going to filter data where value of adults, children and babies = 0

checking_value_0_of_ad_ch_baby = (new_hotel_df['adults']==0) & (new_hotel_df['children']==0) & (new_hotel_df['babies']==0)

# reading data frame where adult,children and babies are 0
new_hotel_df[checking_value_0_of_ad_ch_baby]

hotel  is_canceled  ...  reservation_status  reservation_status_date
2224    Resort Hotel            0  ...           Check-Out               2015-10-06
2409    Resort Hotel            0  ...           Check-Out               2015-10-12
3181    Resort Hotel            0  ...           Check-Out               2015-11-23
3684    Resort Hotel            0  ...           Check-Out               2016-01-04
3708    Resort Hotel            0  ...           Check-Out               2016-01-05
...              ...          ...  ...                 ...                      ...
115029    City Hotel            0  ...           Check-Out               2017-06-30
115091    City Hotel            0  ...           Check-Out               2017-07-01
116251    City Hotel            0  ...           Check-Out               2017-07-17
116534    City Hotel            0  ...           Check-Out               2017-07-22
117087    City Hotel            0  ...           Check-Out               2017-07-29

[180 rows x 30 columns]

**As we know adult, children and babies can't be zero at a time that means these are wrong input so we have to remove this data**

In [ ]:
#reading and storing data where adults,children and babies are having non 0 value 

hotel_df=new_hotel_df[~checking_value_0_of_ad_ch_baby]

In [ ]:
# reading first 5 rows of hotel_df dataframe
hotel_df.head()

In [ ]:
hotel_df['children'].fillna(round(new_hotel_df['children'].mean()), inplace=True)
# Replacing null values of 'children' column with the mean value of the column.

In [ ]:
# Removing rows having empty elements.
hotel_df = new_hotel_df.dropna(axis=0)

In [ ]:

hotel_df.isnull().sum()

# We can see that there are no more null values in the dataset so now the data is cleaned

#**Creating new column**

In [ ]:
# Creating a new column which will show total number of members.
hotel_df['Total_members'] = new_hotel_df['adults'] + new_hotel_df['children'] + new_hotel_df['babies']
hotel_df

In [ ]:
# Creating a new column which will show total number of night stays.
hotel_df['total_night_stays'] = hotel_df['stays_in_weekend_nights']+ hotel_df['stays_in_week_nights']
hotel_df

#**STEP 2 : Analyzing the data**

In [ ]:
Analysis I

#**Count of number of bookings received for a type of hotel**

In [ ]:
## Types of hotel present in the dataset.
hotel_preffered = hotel_df['hotel'].value_counts()
hotel_preffered

There are two types of hotels: 'City hotel' and 'Resort Hotel'.

In [ ]:
# Percentage of hotel bookings accrding to hotel type.
hotel_df['hotel'].value_counts(normalize=True)*100

In [ ]:
# In order to make it easy to understand about the prefference of hotels we can use data visualisation.
hotel_preffered.plot(kind='pie',autopct ='%1.1f%%',figsize =(9,9),fontsize= 16)
plt.title('Bookings wise Hotel types',fontsize = 25)
plt.show()

**Conclusion-** From above graph, we can say city hotels received bookingd that were double of that of resort hotels.

Analysis II

Count of Booking cancellations according to hotel types

In [ ]:
sns.countplot(data=hotel_df, x='hotel', hue='is_canceled')

Conclusion- From above grpah we can observe that number of booking cancellations are higher in City Hotels. Also on comparitive basis cancellation is higher in City Hotels.

Analysis III

Most preffered month for hotel booking

In [ ]:
# Most preffered month by the customers.
most_preffered_month = hotel_df['arrival_date_month'].value_counts()
most_preffered_month

In [ ]:
# Visualisation of data for most preffered month.
most_preffered_month.plot(kind = 'bar',color ='Red',fontsize = 10,figsize=(17,8))
plt.title('MONTH WISE PREFFERENCE',fontsize = 20)
plt.xlabel('MONTH NAME ',fontsize = 16)
plt.ylabel('TOTAL COUNT OF BOOKINGS',fontsize = 16)
plt.show()

# Its clear from this bar chart that most of the bookings received were for the month of August.

Conclusion-  We can observe that guests preferred August month more than other months.

Analysis IV

Country wise guests of hotels


In [ ]:
#Top 15 countries from where guests come 

top_15_countries = hotel_df[hotel_df['is_canceled']==0]['country'].value_counts()[:15]
top_15_countries = top_15_countries.reset_index().rename(columns = {'index':'country','country':'number_of_bookings'})

In [ ]:
#adding percentage column
top_15_countries['percentage'] = (top_15_countries['number_of_bookings']/top_15_countries['number_of_bookings'].sum())*100

In [ ]:
top_15_countries

In [ ]:

sns.barplot(top_15_countries['country'],top_15_countries['percentage']).set(title='Country wise percentage of Guests')


Conclusion-  Large number of guests came from Portugal

Visualizing above information on global map to see if any pattern is present depending on regions.

In [ ]:
import json

In [ ]:
world_map=json.load(open('/content/drive/MyDrive/custom.geo.json',))

In [ ]:
world_map['features'][0]['properties']['adm0_a3']

In [ ]:
import plotly.express as px

In [ ]:
#current plotly version is 4.4.1 which is not compatible for json files.
#We need to upgrade plotly to current version 5.5.0


import plotly
plotly.__version__

In [ ]:
pip install plotly --upgrade

In [ ]:
mapping=px.choropleth(top_15_countries,locations='country',color='number_of_bookings')

In [ ]:
mapping.show()

Conclusion-We can observe that most number of bookings are from Europian countries,USA,Brazil.

To serve current customers of above mentioned countries hotels can improve by offering these guests with curtailed services to improve thier satisfaction.One such thing could be - offering food delicacies famous in those countries.

Developing marketing strategies to attract tourists from asian and african countries can help these hotels in increasing their profits

Which segment brings in the most of the bookings?

In [ ]:
# Different market segments and their total counts in the dataset.
market_segment_booking = hotel_df['market_segment'].value_counts()
market_segment_booking


In [ ]:
# Making a dictionary with these market segment data.
d = dict(market_segment_booking)
print(d)
print(list(d.keys()))
print(list(d.values()))


In [ ]:
# Visualisation market segment data using scatter plot.
plt.figure(figsize=(12,6))
plt.bar(list(dict_1.keys()),list(dict_1.values()))
plt.title('MARKET SEGMENT',fontsize = 20)
plt.ylabel('COUNT',fontsize = 15)
plt.xlabel('UNIQUE MARKET SEGMENTS',fontsize = 15)
plt.grid(True)
plt.show()

Conclusion- We can observe that most bookings received are Online

Analyzing according to meals ordered

In [ ]:
unique_meal_count=new_hotel_df['meal'].value_counts()
unique_meal_count

In [ ]:
# Percentage of meal type count.
percentage_meal_count = new_hotel_df['meal'].value_counts(normalize= True)*100
percentage_meal_count

In [ ]:
# Visualisation of percentage meal count
percentage_meal_count.plot(kind = 'pie',labels=percentage_meal_count.index,figsize=(10,10),autopct='%0.1f%%',fontsize=15)
plt.title('PERCENTAGE OF MEAL PREFERENCE', fontsize=20)
plt.show()

Conclusion-We can observe that most of the guests have booked BB (Bread and Breakfast)

Repeated guests Analysis


In [ ]:
# Count of repeated and non-repeated guest.
print(hotel_df['is_repeated_guest'].value_counts())

# Visualisation.
plt.figure(figsize=(8,8))
sns.countplot(data = hotel_df, x = 'is_repeated_guest')


plt.show()


Conclusion- We can observe that number of repeated guests is very less

Analysis on the basis of ADR(average daily rate)

In [ ]:
# average daily rate according to room type booked
hotel_df=hotel_df[hotel_df['is_canceled']==0]

# Visualisation.
plt.figure(figsize=(12,8))
sns.boxplot(x='assigned_room_type',y='adr',data=hotel_df,hue='hotel')
plt.title('Price of rooms according to room types.')
plt.xlabel('Room_type')
plt.ylabel('Average daily Price')
plt.show()

Conclusion-We can observe that in city hotels highest room ADR was for room type G and was lowest for the room type K. In case of Resort Hotels highest ADR was for room type H and was lowest for room type I.

In [ ]:
# Average ADR of different months and year.
adr_month_year= hotel_df.groupby(['arrival_date_month','arrival_date_year'])['adr'].mean()
adr_month_year.unstack()

In [ ]:

line,ax = plt.subplots(figsize=(18,5)) 
sns.lineplot(x='arrival_date_month',y='adr',data=hotel_df,hue='arrival_date_year',style='arrival_date_year',markers=True,palette='dark',linewidth=5)
ax.set_title('AVERAGE monthly and yearly ADR', fontsize=20)
ax.set_xlabel ('MONTH',fontsize=15)
ax.set_ylabel ('ADR',fontsize=15)
plt.show()

Conclusion- We can observe that average daily count rate has decreasing trend after having peak value in August. This decreasing trend continues till month of January and after month of January ADR starts to increase and this trend is again observed till the month of August. ALso on comaparing year wise we notice that each year ADR has been consistently increasing.

Corralation Plot

In [ ]:
plt.figure(figsize=(20,20))
sns.heatmap(hotel_df.corr(), cmap= 'coolwarm' ,annot=True);
plt.show()

Here dependence of all columns with respect to other columns is represented using heat map.

Comparitively the columns that have relatively higher positive correlation wrt all columns are-

1. lead time and cancellation of booking.
2. adr and presence of children in guests.
3. Previous boooking not cancelled and repeated guest.

Conclusions-

CONCLUSION :-
1. 'City hotels' and 'Resort hotels' are two types of hotels present in the dataset. 'City hotels'are more preffered by the customers than the latter. (66.4% customer prefers 'City hotels' whereas 33.6% customer prefers 'Resort hotels').

2. 'August' witnessed highest number of hotel bookings whereas 'January' witnessed the least.

3. Guests from PRT(Portugal) have got the maximum number of hotel bookings.

4. Its observed that 'City hotels' were more cancelled as compared to 'Resort hotels'.

5. 'Online TA' brings maximum bookings.

6. 'August' has got the highest average ADR in each year.

7. Its found that 77.3% of customers prefers BB(Bread & Breakfast).

8. Its observed that most of the customers rather guests do not repeat their bookings.

9. We can observe that in city hotels highest room ADR was for room type G and was lowest for the room type K. In case of Resort Hotels highest ADR was for room type H and was lowest for room type I.

10. It is observed that average daily count rate has decreasing trend after having peak value in August. This decreasing trend continues till month of January and after month of January ADR starts to increase and this trend is again observed till the month of August.

11. Each year ADR has been increasing consistently.